apis/python/src/tiledbsoma/_read_iters.py

```diff
@@ -5,59 +5,424 @@
+
+    @classmethod
+    def _validate_args(
+        cls,
+        shape: NTuple,
+        axis: Union[int, Sequence[int]],
+        size: Optional[Union[int, Sequence[int]]] = None,
+        reindex_disable: Optional[Union[int, Sequence[int]]] = None,
+    ) -> Tuple[List[int], List[int], List[int]]:
+        """
+        Class method to validate and normalize common user-provided arguments axis, size and reindex_disable.
+        * normalize args to fully specify the arg per dimension, for convenience in later usage
+        * error check and raise if a nonsense value
+        * set defaults where supported.
+        """
+        ndim = len(shape)
+
+        # convert to list
+        axis = list(axis if isinstance(axis, Sequence) else [axis])
+        reindex_disable = (
+            list(
+                reindex_disable
+                if isinstance(reindex_disable, Sequence)
+                else [reindex_disable]
+            )
+            if reindex_disable is not None
+            else []
+        )
+
+        # Currently, only support chunking on one dimension.
+        if len(axis) != 1:
+            raise NotImplementedError(
+                "Multi-dimension blockwise iterators not implemented"
+            )
+        # all dim indices must be in acceptable range
+        if not all(0 <= d < ndim for d in axis):
+            raise ValueError("blockwise `axis` value must be in range [0, ndim)")
+        if not all(0 <= d < ndim for d in reindex_disable):
+            raise ValueError(
+                "blockwise `reindex_disable` value must be in range [0, ndim)"
+            )
+
+        # if not specified, set default size for each axis. Default heuristic
+        # assumes 2D array has many more rows than cols (i.e., n_obs>>n_vars).
+        default_block_size = (2**16,) + (2**8,) * (ndim - 1)
+        if size is None:
+            size = [default_block_size[d] for d in axis]
+        elif isinstance(size, int):
+            size = [size] * len(axis)
+        elif isinstance(size, Sequence):
+            size = list(size) + [default_block_size[d] for d in axis[len(size) :]]
+        else:
+            raise TypeError(
+                "blockwise iterator `size` must be None, int or Sequence[int]"
+            )
+
+        return axis, size, reindex_disable
+
+    @abc.abstractmethod
+    def _create_reader(self) -> Iterator[_RT]:
+        """Sub-class responsibility"""
+        raise NotImplementedError()
+
+    def __next__(self) -> _RT:
+        return next(self._reader)
+
+    def concat(self) -> _RT:
+        """
+        Unimplemented as there is little utility beyond that offered by a ragged
+        read iterator concat, other than reindexing.
+        """
+        raise NotImplementedError("Blockwise iterators do not support concat operation")
+
+    _EagerRT = TypeVar("_EagerRT")
+
+    def _if_eager(
+        self, x: Iterator[_EagerRT], _pool: Optional[ThreadPoolExecutor] = None
```

Comment:
I think the `pool` here is always going to be `self.pool`, right?

Reply:
No.  I want the ability to override the self.pool with the context-managed pool, so it is definitely not the same.  self.pool is whatever the user provided.